### ISSUE TITLE
issue-14 test fails with "global leak detected: addresses"

### ISSUE THREAD

**Issue by BrainCrumbz** [webpack/webpack#3173]
**I'm submitting a bug report**

**Webpack version:**
master branch, commit [80bd8948709c6ffbf93eca6f84ebf6be149d9023](https://github.com/webpack/webpack/commit/80bd8948709c6ffbf93eca6f84ebf6be149d9023)

**Please tell us about your environment:**
Windows 10 x64 version 1511

**Current behavior:**
After setting up a clone of off master, running tests gives:

> ...
> ...
>   6070 passing (8m)
>   32 pending
>   1 failing
> 
>   1) ConfigTestCases extract-text issue-14 should compile:
>      Error: global leak detected: addresses
>       at Runner.checkGlobals (D:\WS\JS\webpack\node_modules\mocha\lib\runner.js:210:21)
>       at Runner.<anonymous> (D:\WS\JS\webpack\node_modules\mocha\lib\runner.js:73:10)
>       at emitOne (events.js:101:20)
>       at Runner.emit (events.js:188:7)
>       at D:\WS\JS\webpack\node_modules\mocha\lib\runner.js:557:14
>       at done (D:\WS\JS\webpack\node_modules\mocha\lib\runnable.js:287:5)
>       at D:\WS\JS\webpack\node_modules\mocha\lib\runnable.js:360:7
>       at _combinedTickCallback (internal/process/next_tick.js:67:7)
>       at process._tickCallback (internal/process/next_tick.js:98:9)

**Expected/desired behavior:**

No test should fail.
- **Browser:** doesn't apply
- **Language:** doesn't apply

**Comment by TheLarkInn** [comment]
I have seen this multiple times and I am not sure what is causing this to fail. Is there anyone skilled in Mocha that can helped track down the global var leak?

**Comment by BrainCrumbz** [comment]
To get a deeper understanding: why do you say "skilled in Mocha"? How's Mocha involved in this?

**Comment by sokra** [comment]
https://github.com/search?utf8=%E2%9C%93&q=addresses+user%3Awebpack&type=Code&ref=searchresults

no idea why this happens...

What's your node.js version?

**Comment by BrainCrumbz** [comment]
* node -v
v6.2.2

* npm -v
3.9.5

* nvm -v
Running version 1.1.1.

**Comment by TheLarkInn** [comment]
I mention this because it is related to a variable leak reported by mocha. After some research I had seen that it was trivial to catch yet I have not had that same experience.

**Comment by BrainCrumbz** [comment]
@TheLarkInn when you say "it was trivial to catch", maybe you're referring to such a hint as mentioned at [this mocha doc page](https://github.com/mochajs/mocha/wiki/Detecting-global-leaks).

@sokra your search prompted me to look for `addresses` under *node_modules* JS/TS files. There are two occurences that seems to miss any variable definition before assignment (so, yer, global usage):

* node_modules\uniqid\examples\browserify\bundle.js, line 254

```
addresses = {};
```

* D:\WS\JS\webpack\node_modules\macaddress\index.js, line 34

```
addresses = {};
```

Might be the second one?

**Comment by BrainCrumbz** [comment]
Yep, that was it. Now, this seems to be an issue with original library, and probably should be just ignored in webpack.
Referenced library version, 0.2.8, still has [this issue](https://github.com/scravy/node-macaddress/blob/0.2.8/index.js#L34) and it's the latest published. [Current master](https://github.com/scravy/node-macaddress/blob/dd079620d11c91957ac897e88d539dfceec41d56/index.js#L37) has the fix, but it's not published yet.

If agreed, solution should be running something like what suggested [here](http://stackoverflow.com/a/10040921/540776).

**Comment by e-cloud** [comment]
`macaddress` isn't required by `webpack`.

According to `yarn.lock`, `css-loader -> cssnano -> postcss-filter-plugins -> uniqid -> macaddress` is the chain.

@BrainCrumbz 's recommended temporary solution is good.

**Comment by TheLarkInn** [comment]
You are suggesting `mocha.setup({globals: ['address']})`?

**Comment by BrainCrumbz** [comment]
`mocha.setup({ globals: ['addresses'] })` or `mocha --globals addresses` from command line (although I'd go for configuration, instead of command line)

**Comment by TheLarkInn** [comment]
Would you like to submit a PR? @sokra should the global leak be tracked in CI? I know we turn it off for CI builds. Or don't use that flag.

**Comment by BrainCrumbz** [comment]
Yep. Although `mocha.setup()` seems to be only for browser environment, so it'll be the command line version

**Comment by BrainCrumbz** [comment]
Hi all.
There was a PR merged to fix this issue. Then recently another PR reverted the change, I guess maybe because original problem has been removed. 

If this problem has gone, anyway, you might want to close this issue.  
HTH

**Comment by najamkhn** [comment]
Seems like its not fixed. Here's all the env details:

```
> git rev-parse --verify HEAD
e15044f3f5c48546f85b76012961aafac5e899b9
> node -v
v6.9.4
> $(npm bin)/mocha --version
3.2.0
```

**Comment by jwalton** [comment]
Appears to be fixed in [macaddress master](https://github.com/scravy/node-macaddress/blob/dd079620d11c91957ac897e88d539dfceec41d56/index.js#L37) since July, but hasn't been pushed to npm yet.  -_-

[uniqid](https://github.com/adamhalasz/uniqid/issues/8) has been made aware of the problem, but hasn't fixed it yet.

Travelling up the dependency tree, I've raised an issue with [postcss-filter-plugins](https://github.com/postcss/postcss-filter-plugins/issues/16).

**Comment by jwalton** [comment]
There's a [closed pull request here](https://github.com/postcss/postcss-filter-plugins/pull/15) for postcss-filter-plugins, unmerged, which would have fixed this.  @ben-eb notes that postcss-filter-plugins will no longer be a dependency of cssnano come v4.

**Comment by jwalton** [comment]
This is now *mostly* fixed in css-loader@0.28.0.  It's fixed provided you don't enable CSS minification in css-loader (which is disabled by default, so it's fixed in the default case at least).

**Comment by ben-eb** [comment]
@jwalton postcss-filter-plugins is not a robust module and caused a lot more problems than it aimed to solve. I'm going to stick with my reasoning from before, which is to deprecate it at the appropriate time and discourage its usage (no big deal as its main dependent is cssnano). I understand that it is a short term pain point but I would rather focus my energy elsewhere to not delay more important fixes which are currently blocking v4.
